libc/string_cpy.c

```diff
@@ -0,0 +1,1482 @@
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(&memStr[10 - testStrLen], &expVal[10 - testStrLen], testStrLen * 2);
+	/* Checking if zeros before and after text are intact */
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(memStr, zeroStr, 6);
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(&memStr[10 + testStrLen], zeroStr, 10);
+
+	/* Copy data in same space but with offset to right */
+	TEST_ASSERT_EQUAL_PTR(&memStr[10 + testStrLen], strncpy(&memStr[10 + testStrLen], &memStr[10], testStrLen));
+	/* Checking if zeros before and after text are intact and text was copied correctly */
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(memStr, expVal, sizeof(memStr));
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_strncpy, one_byte)
+{
+	char buff[CHARS_SET_SIZE] = { 0 },
+		 ascii[CHARS_SET_SIZE] = { 0 };
+	int i;
+
+	/* Copy one by one from created ascii table */
+	for (i = 1; i < CHARS_SET_SIZE; i++) {
+		ascii[i - 1] = i;
+
+		TEST_ASSERT_EQUAL_CHAR_ARRAY(buff, strncpy(buff, &ascii[i], 1), sizeof(buff));
+	}
+}
+
+
+TEST(string_strncpy, various_sizes)
+{
+	/* Pragma truncation disable for the different lengths of copy string tests than source */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+
+	char input[MAX_STR_LEN] = TEST_STR2,
+		 testArray[MAX_STR_LEN] = { 0 },
+		 smallInput[MAX_STR_LEN / 2] = { 0 };
+
+	/* Trying to copy zero bytes */
+	TEST_ASSERT_EQUAL_PTR(testArray, strncpy(testArray, input, 0));
+	TEST_ASSERT_EQUAL_STRING(testArray, "");
+
+	memset(testArray, 0, sizeof(testArray));
+
+	/* Using sizes to copy only part of the array to another */
+	TEST_ASSERT_EQUAL_PTR(testArray, strncpy(testArray, input, sizeof(TEST_STR2) / 2));
+	/* Checking if a copy was executed only on half of the array size */
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(testArray, TEST_STR2, sizeof(TEST_STR2) / 2);
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(&testArray[sizeof(TEST_STR2) / 2], smallInput, sizeof(TEST_STR2) / 2);
+
+	memset(testArray, 0, sizeof(testArray));
+
+	/* Checking ability to stop copy while the size is bigger than a null term that's why sizeof -1 to keep null term*/
+	memset(smallInput, 1, sizeof(smallInput) - 1);
+	TEST_ASSERT_EQUAL_PTR(testArray, strncpy(testArray, smallInput, sizeof(smallInput) * 2));
+	TEST_ASSERT_EQUAL_STRING(testArray, smallInput);
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_strncpy, big)
+{
+	char bigBuff[BIG_NUMB] = { 0 };
+	char *longStr;
+
+	/* Pragma truncation disable for the different lengths of copy string tests than source */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+
+	longStr = testdata_createCharStr(BIG_NUMB);
+	/* Pragma truncation disable for the length of string tests */
+
+	/* Ability to copy long string */
+	TEST_ASSERT_EQUAL_PTR(bigBuff, strncpy(bigBuff, longStr, sizeof(bigBuff)));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(bigBuff, longStr, sizeof(bigBuff));
+
+	free((void *)longStr);
+
+#pragma GCC diagnostic pop
+}
```

Comment:
We can avoid disabling this warning by passing a length of string (not counting NUL term) instead of its size as we have the following note in doc: `bytes that follow a NUL character are not copied`